mavros_extras/src/plugins/distance_sensor.cpp

```diff
@@ -0,0 +1,261 @@
+			auto dist_laser_msg = boost::make_shared<mavros_extras::LaserDistanceSensor>();
+
+			dist_laser_msg->header = header;
+			dist_laser_msg->time_boot_ms = dist_sen.time_boot_ms;
+
+			dist_laser_msg->laser.header = header;
+			dist_laser_msg->laser.angle_min = 0.0;
+			dist_laser_msg->laser.angle_max = 0.0;
+			dist_laser_msg->laser.angle_increment = 0.0;
+			dist_laser_msg->laser.time_increment = 0.0;
+			dist_laser_msg->laser.scan_time = 0.0;
+			dist_laser_msg->laser.range_min = laser_min_range/1E2; 
+			dist_laser_msg->laser.range_max = laser_max_range/1E2;
+			dist_laser_msg->laser.ranges[0] = dist_sen.current_distance/1E2;
+			
+			for (int i = 1; i < dist_laser_msg->laser.ranges.size(); i++) {         
+  				dist_laser_msg->laser.ranges[i] = 0.0;         
+			}
+			for (int f = 0; f < dist_laser_msg->laser.intensities.size(); f++) {     
+				dist_laser_msg->laser.intensities[f] = 0.0;
+			}
+
+			dist_laser_msg->type = dist_sen.type;
+			dist_laser_msg->id = dist_sen.id;
+			dist_laser_msg->orientation = dist_sen.orientation;
+
+			if (dist_laser_msg->covariance != 0.0) dist_laser_msg->covariance = dist_sen.covariance;
+			else dist_laser_msg->covariance = laser_covariance;
+
+			dist_laser_in_pub.publish(dist_laser_msg);
+		}
+
+		else if (dist_sen.type == 1) { // Sonar type
+			auto dist_sonar_msg = boost::make_shared<mavros_extras::SonarDistanceSensor>();
+			
+			dist_sonar_msg->header = header;
+			dist_sonar_msg->time_boot_ms = dist_sen.time_boot_ms;
+			
+			dist_sonar_msg->sonar.header = header;
+			dist_sonar_msg->sonar.radiation_type = sensor_msgs::Range::ULTRASOUND;
+			dist_sonar_msg->sonar.field_of_view = sonar_fov/1E2;
+			dist_sonar_msg->sonar.min_range = sonar_min_range/1E2;
+			dist_sonar_msg->sonar.max_range = sonar_max_range/1E2;
+			dist_sonar_msg->sonar.range = dist_sen.current_distance/1E2;
+			
+			dist_sonar_msg->type = dist_sen.type;
+			dist_sonar_msg->id = dist_sen.id;
+			dist_sonar_msg->orientation = dist_sen.orientation;
+
+			if (dist_sonar_msg->covariance != 0.0) dist_sonar_msg->covariance = dist_sen.covariance;
+			else dist_sonar_msg->covariance = sonar_covariance;
+
+			dist_sonar_in_pub.publish(dist_sonar_msg);
+		}
+	}
+
+	// XXX TODO: Calculate laser/sonar covariances (both in FCU and on the ROS app side)
+
+	// XXX TODO: Determine FOV for Sonar
+
+	/**
+	 * Send laser rangefinder distance sensor data to FCU.
+	 */
+	void send_laser_dist_sensor_data(const ros::Time &stamp, uint16_t min_distance,
+								uint16_t max_distance, uint16_t current_distance) {
+		distance_sensor(stamp.toNSec() / 1000000,
+						min_distance,
+						max_distance,
+						current_distance,
+						0, laser_id,
+						laser_orientation,
+						sonar_covariance);	
+	}
+
+	/**
+	 * Send laser rangefinder distance sensor data to FCU.
+	 */
+	void send_sonar_dist_sensor_data(const ros::Time &stamp, uint16_t min_distance,
+								uint16_t max_distance, uint16_t current_distance) {
+		distance_sensor(stamp.toNSec() / 1000000,
```

Comment:
No, synchronize needed for data comes from FCU. To FCU send current time (in us or ms).